### ISSUE TITLE
Add asserts in PyXXX_GET_SIZE macros

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/cpython#74053]
BPO | [29867](https://bugs.python.org/issue29867)
--- | :---
Nosy | @vstinner, @serhiy-storchaka
PRs | <li>python/cpython#751</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-04-21.23:50:52.034>
created_at = <Date 2017-03-21.07:53:03.456>
labels = ['interpreter-core', 'type-feature', '3.7']
title = 'Add asserts in PyXXX_GET_SIZE macros'
updated_at = <Date 2017-04-21.23:50:52.033>
user = 'https://github.com/serhiy-storchaka'
```

bugs.python.org fields:
```python
activity = <Date 2017-04-21.23:50:52.033>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-04-21.23:50:52.034>
closer = 'vstinner'
components = ['Interpreter Core']
creation = <Date 2017-03-21.07:53:03.456>
creator = 'serhiy.storchaka'
dependencies = []
files = []
hgrepos = []
issue_num = 29867
keywords = []
message_count = 5.0
messages = ['289927', '289929', '290637', '292086', '292089']
nosy_count = 2.0
nosy_names = ['vstinner', 'serhiy.storchaka']
pr_nums = ['751']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue29867'
versions = ['Python 3.7']
```

</p></details>

**Comment by serhiy-storchaka** [comment]
Proposed patch adds asserts for checking the type in macros PyTuple_GET_SIZE, PyList_GET_SIZE and PySet_GET_SIZE. This can help to find the misuse of these macros. Asserts already are used in macros PyBytes_GET_SIZE, PyByteArray_GET_SIZE, PyUnicode_GET_SIZE and PyDict_GET_SIZE. See also the discussion on Python-Dev: https://mail.python.org/pipermail/python-dev/2017-March/147628.html .

This change can break the code that uses these macros for setting the size. For example one place in odictobject.c. But I expect that such cases are rare. And all these macros are not in the limited API.

**Comment by vstinner** [comment]
We should test to run popular C extensions like numpy, PyQt, cython, pillow, etc. with a patched Python to see if it's common to use a "GET" macro to "set" a size. Maybe test also Django see it probably uses many C extensions. I don't ask to test all these modules, but at least a few of them. By "testing", I mean run the test suite of these projects.

**Comment by serhiy-storchaka** [comment]
It would be enough to compile them. New macros produce compiler error when used as lvalue.

But I afraid it will take too much time on my netbook.

**Comment by vstinner** [comment]
New changeset 1a5856bf9295fa73995898d576e0bedf016aee1f by Victor Stinner (Serhiy Storchaka) in branch 'master':
bpo-29867: Add asserts in PyTuple_GET_SIZE, PyList_GET_SIZE and PySet_GET_SIZE. (#751)
https://github.com/python/cpython/commit/1a5856bf9295fa73995898d576e0bedf016aee1f

**Comment by vstinner** [comment]
I merged your PR. We are at the begining of the 3.7 cycle. Applications can be fixed before 3.7.

In the worst case, we can revert the change.

I like this enhancement, it should help to catch bugs ;-)
